### PR TITLE
Prevent iOS App targets from having header build phases

### PR DESCRIPTION
**Pull request by eito**
Resolves https://github.com/tuist/tuist/issues/367

### Short description 📝

This PR prevents the headers build phase from being included in non-framework/library targets.

### Solution 📦

Expose a computed property on `Target` that determines if the headers build phase should be included. In the generation of the headers build phase this new property is checked and only if `true` will the target's headers be unwrapped and subsequently passed to the build phase generation.

### Implementation 👩‍💻👨‍💻

- [x] Reproduce bug based on steps in the issue
- [x] Expose property on `Target` -- `var shouldIncludeHeadersBuildPhase: Bool`
- [x] Change `BuildPhaseGenerator` to include a check for this property before generating headers build phase
- [x] Add test for iOS app targets having an empty headers build phase
- [x] Modify any newly broken tests based on changes (In this case changed `test_generateHeadersBuildPhase_before_generateSourceBuildPhase` to include a non-iOS platform target)

### CODE REVIEW

File: Tests/TuistGeneratorTests/Generator/BuildPhaseGeneratorTests.swift [hunk 1]
```diff
@@ -202,7 +237,8 @@ final class BuildPhaseGeneratorTests: TuistUnitTestCase {
         let headerFileReference = PBXFileReference()
         fileElements.elements[headerPath] = headerFileReference
 
-        let target = Target.test(sources: [(path: "/test/file.swift", compilerFlags: nil)],
+        let target = Target.test(platform: .macOS,
```
Comment (eito) [inline Tests/TuistGeneratorTests/Generator/BuildPhaseGeneratorTests.swift:240]: Changed this from the default (iOS) to macOS because the iOS target shouldn't have the header build phase anymroe.

File: Tests/TuistGeneratorTests/Generator/BuildPhaseGeneratorTests.swift [hunk 2]
```diff
@@ -184,6 +184,41 @@ final class BuildPhaseGeneratorTests: TuistUnitTestCase {
...
+        let sourceFileReference = PBXFileReference(sourceTree: .group, name: "Test")
+        fileElements.elements[path] = sourceFileReference
+
+        let headerPath = AbsolutePath("/test.h")
+        let headers = Headers.test(public: [path], private: [], project: [])
+
+        let headerFileReference = PBXFileReference()
+        fileElements.elements[headerPath] = headerFileReference
+
+        let target = Target.test(platform: .iOS,
```
Comment (eito) [inline Tests/TuistGeneratorTests/Generator/BuildPhaseGeneratorTests.swift:205]: explicitly specifying platform given the test is iOS specific

File: Sources/TuistCore/Models/Target.swift [hunk 3]
```diff
@@ -126,6 +126,16 @@ public struct Target: Equatable, Hashable, Comparable {
         }
     }
 
+    /// Returns true if the target supports having a headers build phase..
+    public var shouldIncludeHeadersBuildPhase: Bool {
+        switch (platform, product) {
+        case (.iOS, .app):
```
Comment (kwridan) [inline Sources/TuistCore/Models/Target.swift:132]: if I'm not mistaken, I believe header build phase is only needed for libraries & frameworks on any platform.

```swift
[
   .framework,
   .staticFramework,
   .staticLibrary,
   .dynamicLibrary
]
```

**Review by kwridan — APPROVED** [review]
Thanks for submitting this @eito 👍 

Left a minor comment, but the changes look good overall. Please update the CHANGELOG when you get a chance.

File: Sources/TuistCore/Models/Target.swift (hunk 3 above)
Comment (pepibumur) [reply Sources/TuistCore/Models/Target.swift:132]: I'd say so.

**Review by pepibumur — APPROVED** [review]
Besides @kwridan's comments, the PR looks great 🚀

**Comment by pepibumur** [comment]
@all-contributosr add @eito for ideas

**Comment by pepibumur** [comment]
@all-contributors add @eito for ideas

**Comment by eito** [comment]
@kwridan just to be clear... you're suggesting I only return true for that explicit subset of product type rather than false for all `.app` types right?

File: Sources/TuistCore/Models/Target.swift (hunk 3 above)
Comment (eito) [reply Sources/TuistCore/Models/Target.swift:132]: updated to only return `true` when one of these product types rather than iOS specific.

**Comment by eito** [comment]
@kwridan @pepibumur I have updated the implementation to account for the 4 supported product types and be platform agnostic. In addition I updated the test `test_generateHeadersBuildPhase_before_generateSourceBuildPhase` so that it tests a `.framework` type instead of the the default `.app` so that it can successfully test the headers build phase before source build phase.

**Review by kwridan — APPROVED** [review]
Awesome thanks for the changes
